Add ability to fetch pod logs as React stream

elvetemedve opened the pull request:
Changelog

- [x] Add ability to stream pod output as React stream

Commit:
Add ability to stream pod output as React stream

Commit:
Fix mocking Guzzle async request

The promise returned by FileHttpClient::asyncRequest() never will be
fulfilled or rejected.
This commit fixes that.

sroze commented on src/Adapter/Http/PodStatusProvider.php:
```diff
@@ -0,0 +1,81 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function setPod(Pod $pod)
```
It looks like to me this should be a constructor's argument

elvetemedve replied:
That was my initial thought, but the provider class could not be reused when the `$pod` instance changes.
I would rather pass the `$pod` argument to the `isTerminated()` and `isPending()` methods, to avoid creating new instances.

sroze commented on src/Adapter/Http/Repository/HttpPodRepository.php:
```diff
@@ -35,13 +38,23 @@ class HttpPodRepository implements PodRepository
+    ) {
         $this->connector = $connector;
         $this->namespaceClient = $namespaceClient;
+        $this->statusProvider = $statusProviderFactory->createFromRepository($this);
```
Not sure of the value of the `StatusProviderFactory` here, if it's just to do a `new [...]` :)

elvetemedve replied:
Well, this is a way to break the circular reference: `PodRepository` needs `PodStatusProvider` and vice versa. Introducing a factory class eliminates the circular dependency.

sroze replied:
~eliminates~ hides 😉

elvetemedve replied:
Call it whatever you like. What I can do about it? :)

elvetemedve replied:
@sroze My proposal is to move the `HttpPodRepository::attach()` and `HttpPodRepository::streamOutput()` methods to the `PodStatusProvider`, because these methods does not really belong to a repository. Then `HttpPodRepository` does not need to have `PodStatusProvider` any more.
But that would be a backward compatibility breaking change, so the `attach()` method should stay for a while (and it would result duplicated code until removed).
What do you think?

sroze commented on src/Adapter/Http/Repository/HttpPodRepository.php:
```diff
@@ -176,12 +189,13 @@ public function exists($name)
     public function attach(Pod $pod, callable $callable)
     {
-        while ($this->isPending($pod)) {
+        $this->statusProvider->setPod($pod);
```
either this `$pod` should be in the constructor of your status provide or as an argument of its `isPending` method.

sroze commented on src/Adapter/Http/Repository/HttpPodRepository.php:
```diff
@@ -195,12 +209,28 @@ public function attach(Pod $pod, callable $callable)
      */
+    public function streamOutput(Pod $pod, LoopInterface $loop): ReadableStreamInterface
+    {
+        $pollIntervalInSeconds = self::ATTACH_LOOP_INTERVAL / 1000;
```
Maybe you can remove the variable here

sroze commented on src/Adapter/React/PodOutputStream.php:
```diff
@@ -0,0 +1,152 @@
+    public function pause()
+    {
+        $this->timer->cancel();
+        if (PromiseInterface::PENDING === $this->httpPromise->getState()) {
```
Could you comment out what's the point of this, I don't get it? Maybe write a spec for this?

elvetemedve replied:
The `pause()` method can be called any time by React when the stream buffer is full and the other side cannot read more data. Because the data comes from an asynchronous Guzzle HTTP call, the promise has to be rejected to stop the streaming. When `pause()` is called right after the promise is fulfilled, calling reject on a fulfilled promise will throw an exception.

I will write spec for this.

elvetemedve replied:
Hmm. Looks like it's not needed any more, because the the wait function is called in the `resume()` method.

sroze commented on src/Model/PodAwareStatusProvider.php:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @package Kubernetes\Client\Repository
+ */
+interface PodAwareStatusProvider
```
I guess you should remove the `setPod` (or move the `$pod` to the method's arguments) and rename this interface `PodStatusProvider`

sroze commented on tests/Adapter/Http/FileHttpClient.php:
```diff
@@ -42,17 +42,19 @@ public function request($method, $path, $body = null, array $options = [])
-            function () use ($filePath) {
+        /** @var PromiseInterface $promise */
+        $promise = new Promise(
+            function () use (&$promise, $filePath) {
```
I don't think you need it to be used as a reference

elvetemedve replied:
I don't think either, but it was used like this in the [Guzzle promise documentation](https://github.com/guzzle/promises#synchronous-wait).

elvetemedve replied:
It must be a limitation of the PHP language. If I remove the `&` operator, I get a notice error:
```
Notice: Undefined variable: promise in tests/Adapter/Http/FileHttpClient.php line 47
```

sroze commented on tests/Adapter/Http/FileHttpClient.php:
```diff
@@ -42,17 +42,19 @@ public function request($method, $path, $body = null, array $options = [])
                 }
 
-                return file_get_contents($filePath);
+                $promise->resolve(file_get_contents($filePath));
```
Actually, that was what the `return` was doing.. not sure why you changed that :)

elvetemedve replied:
Because it did not work. The promise failed with an error like "**Invoking the wait callback did not resolve the promise**".
Maybe a spec is needed for this as well.

Commit:
Update PHPSpec to 4.x

Update the phpspec/phpspec package and fix failing tests.

Commit:
Code refactoring

Commit:
Code refactoring

Move attach() and streamOutput() methods from HttpPodRepository to
HttpPodStatusProvider to eliminate the circular dependency between
these two classes.

sroze commented on src/Adapter/Http/Repository/HttpPodRepository.php:
```diff
@@ -195,12 +203,12 @@ public function attach(Pod $pod, callable $callable)
-     *
-     * @return int
-     */
+       * @param Pod      $pod
```
Any reason for this?

elvetemedve replied:
I blame PHPStorm for it. :) Let me update it quickly.

sroze reviewed (approved):
Like that 👍

sroze commented on src/Adapter/Http/Repository/HttpPodRepository.php:
```diff
@@ -35,11 +36,18 @@ class HttpPodRepository implements PodRepository
-     * @param HttpConnector       $connector
+     * @var PodStatusProvider
+     */
+    private $statusProvider;
```
Is this really useful?

elvetemedve replied:
Good point. That's not used now, as I kept `isTerminated()` and `isPending()` methods for backward compatibility.

sroze replied:
Could you remove it from here then? :)

sroze commented:
Needs a rebase now :)

elvetemedve committed:
Add missing psr/log package

Kubernetes\Client\Adapter\Http\HttpConnector class depends on
Psr\Log\LoggerInterface, but it's not installed by Composer.

elvetemedve committed:
Code refactoring: Adapt to latest React interface

React\EventLoop\LoopInterface::nextTick() method does not exist, use
futureTick() instead.

React\EventLoop\TimerInterface::cancel() method does not exist, use
React\EventLoop\LoopInterface::cancelTimer() instead.

sroze reviewed (approved):
Thank you @elvetemedve!
